### PR TITLE
move envtest helpers into framework package for easier reuse

### PR DESCRIPTION
**Pull request by cheesesashimi**
**- What I did**

I moved the EnvTest setup and environment cleanup code into the `test/framework` package for easier reuse across the rest of the MCO codebase. Additionally, I refactored the Makefile as well as some of the test scripts to collect JUnit test reports in CI as well as a test coverage report.

**- How to verify it**

Because `test/e2e-bootstrap` is the only test suite that uses EnvTest, one can run `$ make bootstrap-e2e` or run the test target via `/test bootstrap-unit`. (For some unknown reason, we're not running `bootstrap-unit` on each PR, though it makes sense to because they're not overly resource intensive and execute quickly.)

For the JUnit and test coverage reports, open the links for the various Prow CI jobs and you should see them enumerated in the Test Results pane as well as a red / green graphic illustrating test coverage.

**- Description for the changelog**
Move EnvTest helpers into framework package for easier reuse

### CODE REVIEW

**Comment by cheesesashimi** [comment]
Converting to draft to conserve CI resources while I debug the bootstrap-unit stage.

**Comment by cheesesashimi** [comment]
This is now ready for review!

**Review by yuqi-zhang — APPROVED** [review]
Generally lgtm! Tested locally as well just to make sure and all the tests seem to run as expected.

Will let others give a look as well before lgtm

File: Makefile [hunk 1]
```diff
@@ -64,14 +78,42 @@ go-deps:
 	chmod +x ./vendor/k8s.io/code-generator/generate-groups.sh
 	chmod +x ./vendor/k8s.io/code-generator/generate-internal-groups.sh
 
-install-tools:
```
Comment (cgwalters) [inline Makefile:67]: I started to dig into writing this like:

```
TOOLS := setup-envtest go-junit-report golangci-lint
TOOLS_NEEDED := $(foreach tool,$(TOOLS),$(shell command -v go-junit-report > /dev/null)
```

But...honestly it's probably not worth the Make wizardary at this scale.  If we had 3 more tools to install it'd be worth reducing the copy paste, but IMO not yet.

File: hack/golangci-lint.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+set -o nounset
+set -o pipefail
```
Comment (cgwalters) [inline hack/golangci-lint.sh:4]: I'd still do `set -euo pipefail`, then below:

File: hack/golangci-lint.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+if [ ! -n "$GOTAGS" ]; then
+  echo "No Go tags provided"
+  exit 1
+fi
+
+cd "$REPO_ROOT"
+
+golangci-lint run --build-tags="$GOTAGS";
+retval="$?"
```
Comment (cgwalters) [inline hack/golangci-lint.sh:24]: The bash equivalent of "try/catch" I've seen is:
```
retval=0
golangci-lint run --build-tags="$GOTAGS" || retval=$?
```

Notably this continues to work with `set -e`.

Comment (cheesesashimi) [reply hack/golangci-lint.sh]: Ahh thanks for the tip! I've seen (and done) this before, but couldn't remember how it interacts with `set -e`. I've incorporated those changes nevertheless.

File: Makefile (hunk 1 above)
Comment (cheesesashimi) [reply Makefile:67]: Yeah, I wasn't sure what the best pattern would be for this.

**Pull request merged**